Fix bugs in RestrictedPerm

ChrisJefferson opened the pull request:
The code in RestrictedPerm for handling ranges had a few bugs (I found issue (2) in some code I was working on, when I went to investigate I noticed issue (1))

1) If the list contained negative numbers, () was always returned
2) Ranges with steps which were not step = 1 were mishandled -- but (I believe) this would never return a wrong answer, just an Error about invalid input, when the input was value.

Rather than try to fix the old code, which was both buggy and confusing, I just wrote a (hopefully) much simpler loop, which I believe will be just as efficient (as the only one wasn't trying to do "clever skipping" anyway.

wilfwilson commented on src/permutat.cc:
```diff
@@ -1748,19 +1748,10 @@ static inline Obj RESTRICTED_PERM(Obj perm, Obj dom, Obj test)
-      while (p<=i) {
-        ptRest[p]=ptPerm[p];
-        p+=inc;
+      for ( i = p; i != p+inc*len; i += inc) {
+          if(i >= 1 && i <= deg) {
```
You could combine these if you wanted to into something like:
```suggestion
      for ( i = Maximum(1,p); i != p+inc*len && i <= deg; i += inc) {
```
(Not sure how you do `Maximum`)

ChrisJefferson replied:
Be aware that inc can be positive or negative, and p could start higher than deg (that's why I do i != p+inc*len, rather than the more usual i < p+inc*len).

wilfwilson replied:
Oops, silly me, you're quite right. Please continuing ignoring my wrong suggestion ;-)

wilfwilson reviewed (approved):
Good catch.

ChrisJefferson commented:
Just to say, we could also change RestrictedPerm to just reject inputs which contain non-positive integers (it doesn't really make sense to restrict a perm on negative inputs), but that would be a change from current behaviour on non-ranges (which don't currently have any bugs I know of).

fingolfin commented on src/permutat.cc:
```diff
@@ -1748,19 +1748,10 @@ static inline Obj RESTRICTED_PERM(Obj perm, Obj dom, Obj test)
-      while (p<=i) {
-        ptRest[p]=ptPerm[p];
-        p+=inc;
+      for ( i = p; i != p+inc*len; i += inc) {
+          if(i >= 1 && i <= deg) {
```
```suggestion
      for (i = p; i != p + inc * len; i += inc) {
          if (i >= 1 && i <= deg) {
```

ChrisJefferson replied:
fixed by clang-formatting, thanks

ChrisJefferson commented:
Code now reformatted and simplified somewhere, I now (correctly) just reject negative inputs. Added (possibly too many) tests, to make sure all the various boundary conditions are hit properly.

wilfwilson reviewed (approved):
Great.

fingolfin commented:
Backported to stable-4.11 via PR #4262
